### DISCUSSION TITLE
How can I create a generic middleware function, that is not attached to a specific router?

### DISCUSSION THREAD

**Discussion by moltar** [trpc/trpc#4092] [Q&A]
I'd like to be able to compose the router from a few generic middleware fns. But the types for them are so gnarly, I am at a loss.

How can I have a simple function definition for a middleware, that can be reused in many routers?

Also, asked in Discord: https://discord.com/channels/867764511159091230/1088119988323360779

**Comment by kamilogorek** [comment]
Can you provide a simple example where it's not working? As Nick already pointed out on Discord, middleware functions are not tied in any way to the input/output, they are just a passthrough, so you should be able to reuse them just fine.

**Reply by moltar** [reply]
Sure, here you go: [Playground Link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzsAdsGAVASgBQMJwC+cAZlBCHAOQACMUYAxgPQDOAplAG6dUDcAKAGMIKVvHgBeZGky48AOkZR2AQxjsAFAgFw4zZnEC8G4HEdwDwbgb-24IkCHYoJAC2Cs4EAK4SIcEKoDW7HAwjoGc5FBwAObeqgDuqgCeunBh0ABi0L4wGlAAXHDacKyOqmDsADQpUOFEAJRwkgB8iMl6KjDuUCgter1wCgPFpRWtvalQ5aOEgnqEk4S1ggLsAB6QsNai4nClYFgeOQ1BCuSenNrJGuL5MApg5IzsACad7KMK7hyao3oGxuZWGAJMpVcI-HyxTS1SZ9WrvACO7k4CU0UIazQQC3miyEJHcKEYMGAoghaJ0ehEYngSBAwCeTwANux4ioiEdUOhsPglKINCsYFClCp1Foccl2p1urT6UyWVpCihVjBKoSVnV0XBFfzCqq6jjCAIgA)

---

```ts
import { initTRPC } from '@trpc/server';

const t = initTRPC.create({
  // 👇🏻 comment this out to make the error go away
  errorFormatter: ({ shape, error }) => {
    return {
      ...shape,
      error,
    };
  },
});

export const appRouter = t.router({
  test: t.procedure
    .use(
      // 👇🏻 type error
      mw(),
    )
    .query(() => {}),
});

function mw() {
  const { middleware } = initTRPC.context().create();

  return middleware(({ next, ctx }) => next({ ctx }));
}
```

**Reply by kamilogorek** [reply]
Ah, this might be more complicated though, as `MiddlewareFunction` type is tied to the return type of `TRPCBuilder` (which is what `initTRPC.create` returns. There are some ideas at https://github.com/trpc/trpc/issues/3434#issuecomment-1363774321 but not sure if that'll cover all your cases.

**Comment by Nick-Lucas** [comment]
There is a lot of merit to this need. It's common for libraries like auth providers create to provide out-the-box middlewares which you can just configure and plug in to a web server, but currently you even just need the Context to be identical in order to plug a middleware straight in:

<img width="584" alt="image" src="https://user-images.githubusercontent.com/8896153/228356121-d955591a-9586-43ec-8266-43726dc40f7d.png">

I think this could be solved with a factory pattern, albeit we would need to create an `AnyTRPC` type in a better way than the below:

<img width="536" alt="image" src="https://user-images.githubusercontent.com/8896153/228357689-53f942c5-a1e9-4c1a-8bb7-e2af5605b0f1.png">

**Reply by Nick-Lucas** [reply]
The trouble with the above is adding an errorFormatter or transformer breaks it, but this would be solved with a proper AnyTRPC type which can be used for the generic and covers _all_ keys.

The next problem will be "I would like the Context to have a particular key like req/res on it" or "I would like this key from Meta" or "I would like to extend/swap the Context". These are probably all solvable in the same way, just not easy to do right now.
